Readme.md
This OPC implementation is based on free prOpc implementation of "Production Robots Engineering Ltd" http://www.prel.co.uk and is extended by new features.

OLE for Process Control or 'OPC' is a well established standard for communication between process plant and software running under Windows. The specifications are designed and maintained by the OPC Foundation a non-profit making organisation founded by many of the leading players in the process control industry.

Programming OPC clients and servers can be a daunting task, especially for those with little or no experience of COM programming. We think you will be surprised how simple OPC server programming can be with the prOpc toolkit; why not download a copy now.

OPC server supports DA1.0, DA2.05 and DA3.0 interfaces and passes OPC Compliance Tests.

**prOpcKit Features:**

  * _OPC Server Wizard_
> By filling in a simple form you can have an OPC data access server up and running in a couple of minutes. The OPC server wizard writes a simple class definition: all you have to do is add the published properties you want visible to your OPC clients. For a quick introduction to the wizard see An Introduction to the Opc Server Wizard.
  * _TOpcSimpleClient component_
> Turn your Delphi application into an OPC client by dropping the TOpcSimpleClient component onto one of your forms.
  * _Build an Instant Client with the 'Very Simple Client' unit_
> Build yourself a moderately serviceable OPC Data Access client in a single line of code. 70% of the functionality for 30% of the work. For a brief explanation of how this works see Using the Very Simple Client.
  * _Complete On-line and tutorial documentation_
> There is a complete manual explaining how to build OPC programs in the simplest possible terms. In addition, there is complete on-line reference documentation in .hlp format, fully compatible with Delphi's OpenHelp system.
  * _Numerous Examples_
> There are a number of example servers and clients provided with the kit.
  * _100% Delphi_
> The prOpcKit libraries are native Delphi. There are no external DLLs.
  * _High Performance_
> All prOpcKit servers and clients based on TOpcSimpleClient use the OPC custom interfaces directly, offering the same performance as C++ implementations.

ReleaseNotes.txt
{The MIT License (MIT)
 prOpc Toolkit
 Copyright (c) 2000, 2001 Production Robots Engineering Ltd

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.}

v1.16e Maintenance Release 08/11/07
***********************************
Use OPCENUM to when opening TOpcSimpleClient unless at design time.


v1.16d Maintenance Release 08/11/07
***********************************
Fixed bug in handing boolean arrays in prOpcVarUtils.


v1.16c Maintenance Release 29/09/07
***********************************
1.16c.1 Added new build to support "Delphi 2007 for Win32". Setup program is 
D11Setup.exe.


v1.16b Maintenance Release 08/04/07
***********************************
1.16b.1 prOpcEnum.pas
Fixed memory leak in function OpenReg when failed to connect to remote registry.

1.16b.2 prOpcEnum.pas
Added new (optional) parameter to GetServerInfo: Options: TServerSearchOptions.
if this includes soUseOpcEnum then the Opc enumerator service (OpcEnum.exe) is
used to get the information, rather than remote registry access. Note that if
this option is used, then the 'Vendor' parameter always returns empty, as this
information is not available using OpcEnum.


v 1.16a Maintenance Release 06/02/07
************************************
1.16a.1 Added basic support for Async write to SimpleClient component. Server must
support IAsyncIO2.


v 1.16 Upgrade Release 10/02/06
*******************************
Support for Delphi 2005 and Delphi 2006. The Delphi 2005 build is included for customer's
convenience but Delphi 2005 is not an officially supported platform for prOpcKit. 


v 1.15h Maintenance Release 30/06/05
************************************
1.15.1 prOpcClient.pas
Modify 'TOpcGroup.DisconnectServer' to avoid throwing exception if connection point cannot
be found (e.g Server has unexpectedly died).


v 1.15g Maintenance Release 27/01/05
************************************
1.15.6.1    TItemIdList now ignores duplicates to allow InvalidateNamespace with
            'keep existing'. This was a bug.

1.15.6.2    Fixed AV in InvalidateNamespace



v 1.15f Maintenance Release 15/06/04
************************************
1.15f.1 prOpcClient.pas
Added property TOpcGroup.Timebias. This is initialised from the 
server when the group is connected.

1.15f.2
Added property TOpcGroup.GroupInterface to allow access to underlying
OpcInterfaces.


v 1.15e Maintenance Release 03/04/04
************************************
Added "ForceDisconnect" function to Group and ClientInfo to allow 
disfunctional clients to be disconnected from the server

Fixed bug in TAsyncTask - ActiveItemList in TRefreshTask was being
freed before the Task was processed. This was quite a nasty bug
but there have been no reported issues from the field traceable to 
this problem - it was detected while carrying out routine testing
of the implementation of Forced disconnect.


v 1.15d Maintenance Release 15/11/03
************************************
Minor update now includes prOpcKit.dcp in distribution.


v 1.15c Maintenance Release 17/10/03
************************************
Version resource rationalised. This is now marked 1.15.3 rather 
than 1.1.5.0 which was a bit obscure, even to me.

Bug in prOpcRttiServer 
EUInfo was returned for Boolean items which is correct for Native
Boolean but incorrect for Variant Bool since true = -1. This was
causing an AV in the latest release of Terravic tools - which now
support EUInfo. Which is nice providing one does not have bugs in 
one's implementation.

Slight change to TServerSelectDlg.Execute - 'Options' parameter
added to allow use of OpcEnum.


v 1.15b Maintenance Release 19/03/03
***********************************
* Fixed bug in prOpcServer.pas. Memory Leak in IOPCAyncIO2.Refresh2.
* Fixed broken example ClientDemo3


v 1.15 Maintenance Release 20/09/02
***********************************
* Support for D7
* Fixed broken example prDemo5


v 1.14 Upgrade Release 14/04/02
*******************************
 * Detailed release notes now found in prOpcKit.doc.
 * Added support for hierarchical browsing 
 * Removed Rtti code from prOpcServer.pas and moved it to another unit. This 
   will require small source code changes to servers that use Rtti.
 * Changed TOpcSimpleClient so that Items in groups are no longer sorted by 
   name. This was very eccentric behaviour, but the change may have an impact
   on existing clients. See prOpcKit.doc.
 * Added facility to allow servers to detect and act on COM errors when making
   callbacks to clients.
 * Minor bug fixes to prOpcServer. 
 * Evaluation Edition no longer links Masks.dcu and ignores szFilterCriteria 
   parameter to IOPCServerBrowseServerAddressSpace.BrowseOPCItemIDs.
 * Includes Very Simple Client prOpcVSC.
 * Memory leak in TOpcGroup.SyncRead (and possibly others) fixed.

v 1.13a Maintenance Release 10/09/01
************************************

Sources/prOpcUtils.pas
{------------------------------------------------------------}
{The MIT License (MIT)

 prOpc Toolkit
 Copyright (c) 2000, 2001 Production Robots Engineering Ltd

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.}
{------------------------------------------------------------}
unit prOpcUtils;
{$I prOpcCompilerDirectives.inc}

{History
  13-03-01. Dropped PathChar property. I can see real problems with this,
            and I think it is very unlikely to be ever needed.
  14-03-10. Numerous mods to handle new exception scheme.


Release 1.14
1.14.1      Fixed a few memory leaks caused by passing structures by value (!).
            I lecture everyone endlessly about this and then get caught out
            myself! This bug may be an issue at RR Derby.
}

{do not remove the following comment}

//CE_Desc_Include(..\Help\Topics.txt)

interface
uses
  SysUtils, Windows, ActiveX, Classes, prOpcDa, prOpcComn, prOpcTypes,
  prOpcError;

type
  DAVariant = array of OleVariant;

procedure FreeOPCItemState(var v: OPCITEMSTATE);
procedure FreeOPCItemStateArray(dwCount: UINT; pv: POPCITEMSTATEARRAY);

procedure FreeOPCItemResult(var v: OPCITEMRESULT);
procedure FreeOPCItemResultArray(dwCount: UINT; pv: POPCITEMRESULTARRAY);

procedure FreeOPCItemAttributes(var v: OPCITEMATTRIBUTES);
procedure FreeOPCItemAttributesArray(dwCount: UINT; pv: POPCITEMATTRIBUTESARRAY);

procedure FreeOPCItemProperties(var v: OPCITEMPROPERTIES);

procedure FreeOPCItemBrowseElement(v: OPCBROWSEELEMENT);
procedure FreeOPCItemBrowseElementArray(dwCount: UINT; pv: POPCBROWSEELEMENTARRAY);

procedure FreeOleVariantArray(dwCount: Integer; pv: POleVariantArray);

procedure FreeResultList(pv: PResultList);

procedure FreeServerStatus(pv: POPCSERVERSTATUS);

procedure FormatOPCStmDataTime(Stream: Pointer; var GroupHeader: POPCGROUPHEADER;
                                var ItemHeaders: POPCITEMHEADER1ARRAY; var Data: DAVariant);
procedure FormatOPCStmData(Stream: Pointer; var GroupHeader: POPCGROUPHEADER;
                            var ItemHeaders: POPCITEMHEADER2ARRAY; var Data: DAVariant);
procedure FormatOPCStmWriteComplete(Stream: Pointer; var GroupHeader: POPCGROUPHEADERWRITE;
                                     var ItemHeaders: POPCITEMHEADERWRITEARRAY);

function GetMalloc: IMalloc;

function GIT : IGlobalInterfaceTable;

function DatatypeToStr(Datatype: Integer): string;
function QualityToStr(Quality: Word): String;
function AccessRightsToStr(AccessRights: DWORD): String;
procedure InitialiseClientSecurity;

function GetOpcErrorString(Server: IUnknown; Code: HRESULT): string;

implementation
uses
{$IFDEF D6UP}
  Variants,
{$ENDIF}
  ComObj;

resourcestring
  SUnknownError = 'Unknown error code %.8x';

const
  CLSID_StdGlobalInterfaceTable : TGUID = '{00000323-0000-0000-C000-000000000046}';

var
  cGIT : IGlobalInterfaceTable = nil;

function GIT : IGlobalInterfaceTable;
begin
  if (cGIT = nil) then
    OleCheck(CoCreateInstance(CLSID_StdGlobalInterfaceTable, nil, CLSCTX_ALL,
      IGlobalInterfaceTable, cGIT));
  Result := cGIT;
end;

function GetMalloc: IMalloc;
begin
  OleCheck(CoGetMalloc(1, Result))
end;

procedure FreeOPCItemState( var v: OPCITEMSTATE);
begin
  v.vDataValue:= Unassigned
end;

procedure FreeOPCItemStateArray( dwCount: UINT; pv: POPCITEMSTATEARRAY);
var
  i: Integer;
begin
  if Assigned(pv) then
  begin
    for i:= 0 to dwCount - 1 do
      FreeOPCItemState(pv^[i]);
    CoTaskMemFree(pv)
  end
end;

procedure FreeOPCItemResult( var v: OPCITEMRESULT);
begin
  with v do if Assigned(pBlob) then
    CoTaskMemFree(pBlob)
end;

procedure FreeOPCItemResultArray( dwCount: UINT; pv: POPCITEMRESULTARRAY);
var
  i: Integer;
begin
  if Assigned(pv) then
  begin
    for i:= 0 to dwCount - 1 do
      FreeOPCItemResult(pv^[i]);
    CoTaskMemFree(pv)
  end
end;

procedure FreeOPCItemAttributes( var v: OPCITEMATTRIBUTES);
begin
  with v do
  with GetMalloc do
  begin
    Free(szAccessPath);
    Free(szItemID);
    Free(pBlob);
    vEUInfo:= Unassigned
  end
end;

procedure FreeOPCItemAttributesArray( dwCount: UINT; pv: POPCITEMATTRIBUTESARRAY);
var
  i: Integer;
begin
  if Assigned(pv) then
  begin
    for i:= 0 to dwCount - 1 do
      FreeOPCItemAttributes(pv^[i]);
    CoTaskMemFree(pv)
  end
end;

procedure FreeOPCItemProperties(var v: OPCITEMPROPERTIES);
var
  i: Integer;
begin
  with v do
  with GetMalloc do
  begin
    if dwNumProperties > 0 then
      for i:= 0 to dwNumProperties - 1 do
      begin
        Free(pItemProperties[i].szItemID);
        Free(pItemProperties[i].szDescription);
        pItemProperties[i].vValue := Unassigned;
      end;
    Free(pItemProperties);
  end;
end;

procedure FreeOPCItemBrowseElement(v: OPCBROWSEELEMENT);
begin
  with v do
  with GetMalloc do
  begin
    Free(szName);
    Free(szItemID);
    FreeOPCItemProperties(ItemProperties);
  end
end;

procedure FreeOPCItemBrowseElementArray(dwCount: UINT; pv: POPCBROWSEELEMENTARRAY);
var
  i: Integer;
begin
  if Assigned(pv) then
  begin
    if dwCount > 0 then
      for i:= 0 to dwCount - 1 do
        FreeOPCItemBrowseElement(pv^[i]);
    CoTaskMemFree(pv)
  end
end;

procedure FreeOleVariantArray( dwCount: Integer; pv: POleVariantArray);
var
  i: Integer;
begin
  if Assigned(pv) then
  begin
    for i:= 0 to dwCount - 1 do
      pv^[i]:= Unassigned;
    CoTaskMemFree(pv)
  end
end;

procedure FreeResultList(pv: PResultList);
begin
  if Assigned(pv) then
    CoTaskMemFree(pv)
end;

procedure FreeServerStatus(pv: POPCSERVERSTATUS);
begin
  CoTaskMemFree(pv.szVendorInfo);
  CoTaskMemFree(pv);
end;

type
  TDataPtr = record
  case Byte of
    0: (Data: POleVariant);
    1: (StrLen: PDWord);
    2: (OleStr: PWideChar);
    3: (VarArray: PVarArray);
  end;

procedure ReadVariant(Stream: Integer;  var Data: OleVariant);
var
  DataPtr: TDataPtr absolute Stream;
  StmData: POleVariant;
begin
  StmData:= DataPtr.Data;
  with TVarData(StmData^) do
  case VType of
    varOleStr:
    begin
      Inc(Stream, SizeOf(TVarData) + SizeOf(DWORD));
      VOleStr:= DataPtr.OleStr
    end;
    varArray:    {I have been unable to test this}
    begin
      Inc(Stream, SizeOf(TVarData));
      VArray:= DataPtr.VarArray;
    end
  end;
  Data:= StmData^
end;

procedure FormatOPCStmDataTime( Stream: Pointer; var GroupHeader: POPCGROUPHEADER;
                                var ItemHeaders: POPCITEMHEADER1ARRAY; var Data: DAVariant);
var
  StmInt: DWORD absolute Stream;
  i: Integer;
begin
  GroupHeader:= Stream;
  ItemHeaders:= Pointer(StmInt + SizeOf(OPCGROUPHEADER));
  SetLength(Data, GroupHeader^.dwItemCount);
  for i:= 0 to GroupHeader^.dwItemCount - 1 do
    ReadVariant(StmInt + ItemHeaders^[i].dwValueOffset, Data[i])
end;

procedure FormatOPCStmData( Stream: Pointer; var GroupHeader: POPCGROUPHEADER;
                            var ItemHeaders: POPCITEMHEADER2ARRAY; var Data: DAVariant);
var
  StmInt: DWORD absolute Stream;
  i: Integer;
begin
  GroupHeader:= Stream;
  ItemHeaders:= Pointer(StmInt + SizeOf(OPCGROUPHEADER));
  SetLength(Data, GroupHeader^.dwItemCount);
  for i:= 0 to GroupHeader^.dwItemCount - 1 do
    ReadVariant(StmInt + ItemHeaders^[i].dwValueOffset, Data[i])
end;

procedure FormatOPCStmWriteComplete( Stream: Pointer; var GroupHeader: POPCGROUPHEADERWRITE;
                                     var ItemHeaders: POPCITEMHEADERWRITEARRAY);
var
  StmInt: Integer absolute Stream;
begin
  GroupHeader:= Stream;
  inc(StmInt, SizeOf(OPCGROUPHEADER));
  ItemHeaders:= Stream
end;

function DatatypeToStr(Datatype: Integer): string;
begin
  case Datatype of
    varEmpty: Result:= 'Empty';
    varNull: Result:= 'Null';
    varSmallint: Result:= 'Smallint';
    varInteger: Result:= 'Integer';
    varSingle: Result:= 'Single';
    varDouble: Result:= 'Double';
    varCurrency: Result:= 'Currency';
    varDate: Result:= 'Date';
    varOleStr: Result:= 'OleStr';
    varDispatch: Result:= 'Dispatch';
    varError: Result:= 'Error';
    varBoolean: Result:= 'Boolean';
    varVariant: Result:= 'Variant';
    varShortInt: Result := 'ShortInt';
    varByte: Result := 'Byte';
    varWord: Result := 'Word';
    varLongWord: Result := 'LongWord';
    varInt64: Result := 'Int64';
  else
    Result:= 'Unknown';
  end
end;

function QualityToStr(Quality: Word): String;
begin
  case (Quality and OPC_QUALITY_MASK) of
    OPC_QUALITY_BAD:
    begin
      Result:= 'Bad ';
      case Quality and OPC_STATUS_MASK of
        OPC_QUALITY_CONFIG_ERROR: Result:= Result + 'Config error ';
        OPC_QUALITY_NOT_CONNECTED: Result:= Result + 'not connected ';
        OPC_QUALITY_DEVICE_FAILURE: Result:= Result + 'device failure ';
        OPC_QUALITY_SENSOR_FAILURE: Result:= Result + 'sensor failure ';
        OPC_QUALITY_LAST_KNOWN: Result:= Result + 'last known ';
        OPC_QUALITY_COMM_FAILURE: Result:= Result + 'comm failure ';
        OPC_QUALITY_OUT_OF_SERVICE: Result:= Result + 'out of service ';
      end
    end;
    OPC_QUALITY_UNCERTAIN:
    begin
      Result:= 'Uncertain ';
      case Quality and OPC_STATUS_MASK of
        OPC_QUALITY_LAST_USABLE: Result:= Result + 'last usable ';
        OPC_QUALITY_SENSOR_CAL: Result:= Result + 'sensor cal ';
        OPC_QUALITY_EGU_EXCEEDED: Result:= Result + 'egu exceeded ';
        OPC_QUALITY_SUB_NORMAL: Result:= Result + 'sub normal ';
      end
    end;
    OPC_QUALITY_GOOD:
    begin
      Result:= 'Good ';
      case Quality and OPC_STATUS_MASK of
        OPC_QUALITY_LOCAL_OVERRIDE: Result:= Result + 'local override ';
      end
    end;
  end;
  case Quality and OPC_LIMIT_MASK of
    OPC_LIMIT_OK: Result:= Result + 'limit ok';
    OPC_LIMIT_LOW: Result:= Result + 'limit low';
    OPC_LIMIT_HIGH: Result:= Result + 'limit high';
    OPC_LIMIT_CONST: Result:= Result + 'limit const';
  end
end;

function AccessRightsToStr(AccessRights: DWORD): String;
begin
  Result:= '';
  if (AccessRights and OPC_READABLE) <> 0 then
    Result:= Result + 'r';
  if (AccessRights and OPC_WRITABLE) <> 0 then
    Result:= Result + 'w'
end;


procedure InitialiseClientSecurity;
const
  RPC_C_AUTHN_LEVEL_NONE = 1;
  RPC_C_IMP_LEVEL_IMPERSONATE = 3;
  EOAC_NONE = 0;
var
  Res: HRESULT;
begin
  Res:= CoInitializeSecurity(nil, -1, nil, nil,
    RPC_C_AUTHN_LEVEL_NONE,
    RPC_C_IMP_LEVEL_IMPERSONATE, nil, EOAC_NONE, nil);
  if (Res <> RPC_E_TOO_LATE) then
    OleCheck(Res)
end;

function GetOpcErrorString(Server: IUnknown; Code: HRESULT): string;
{this function is for clients...}
var
  Common: IOPCCommon;
  ppString: PWideChar;
  Buf: array[0..255] of Char;
begin
  if not StdOpcErrorToStr(Code, Result) then
  begin
    if Assigned(Server) and  {ask the server}
       (Server.QueryInterface(IOPCCommon, Common) = S_OK) and
       (Common.GetErrorString(Code, ppString) = S_OK) then
    begin
      Result:= ppString;
      CoTaskMemFree(ppString)
    end else
    if FormatMessage(FORMAT_MESSAGE_FROM_SYSTEM or FORMAT_MESSAGE_ARGUMENT_ARRAY,
          nil, DWORD(Code), 0, Buf, SizeOf(Buf), nil) > 0 then
    begin
      Result:= Buf
    end else
    begin  {give up}
      FmtStr(Result, SUnknownError, [Code])
    end
  end
end;

end.
